cpp/tests/groupby/groupby_test_util.hpp

```diff
@@ -84,26 +99,73 @@ inline void test_single_agg(column_view const& keys,
-    cudf::test::detail::expect_columns_equivalent(
-      expect_vals, *result.second[0].results[0], debug_output_level::ALL_ERRORS);
+  if (use_sort == force_use_sort_impl::YES && keys_are_sorted == sorted::NO) {
+    CUDF_TEST_EXPECT_TABLES_EQUAL(*sorted_expect_keys, result.first->view());
+    cudf::test::detail::expect_columns_equivalent(sorted_expect_vals->get_column(0),
+                                                  *result.second[0].results[0],
+                                                  debug_output_level::ALL_ERRORS);
+
   } else {
-    auto const sort_order  = sorted_order(result.first->view(), {}, {null_order::AFTER});
+    auto const sort_order  = sorted_order(result.first->view(), column_order, precedence);
     auto const sorted_keys = gather(result.first->view(), *sort_order);
     auto const sorted_vals = gather(table_view({result.second[0].results[0]->view()}), *sort_order);
 
-    CUDF_TEST_EXPECT_TABLES_EQUAL(table_view({expect_keys}), *sorted_keys);
-    cudf::test::detail::expect_columns_equivalent(
-      expect_vals, sorted_vals->get_column(0), debug_output_level::ALL_ERRORS);
+    CUDF_TEST_EXPECT_TABLES_EQUAL(*sorted_expect_keys, *sorted_keys);
+    cudf::test::detail::expect_columns_equivalent(sorted_expect_vals->get_column(0),
+                                                  sorted_vals->get_column(0),
+                                                  debug_output_level::ALL_ERRORS);
   }
 }
 
+inline auto sum_agg() { return cudf::make_sum_aggregation<groupby_aggregation>(); }
+
+inline void test_hash_based_sum_agg(column_view const& keys,
+                                    column_view const& values,
+                                    column_view const& expect_keys,
+                                    column_view const& expect_vals)
+{
+  test_single_agg(keys,
+                  values,
+                  expect_keys,
+                  expect_vals,
+                  sum_agg(),
+                  force_use_sort_impl::NO,
+                  null_policy::INCLUDE);
+}
+
+inline void test_sort_based_sum_agg(column_view const& keys,
+                                    column_view const& values,
+                                    column_view const& expect_keys,
+                                    column_view const& expect_vals)
+{
+  test_single_agg(keys,
+                  values,
+                  expect_keys,
+                  expect_vals,
+                  sum_agg(),
+                  force_use_sort_impl::YES,
+                  null_policy::INCLUDE);
+}
+
+inline void test_sum_agg(column_view const& keys,
+                         column_view const& values,
+                         column_view const& expected_keys,
+                         column_view const& expected_values)
+{
+  test_sort_based_sum_agg(keys, values, expected_keys, expected_values);
+  test_hash_based_sum_agg(keys, values, expected_keys, expected_values);
```

Comment:
I think this can help reducing code duplicates:
```
  auto const do_test = [&](auto const use_sort_option) {
    test_single_agg(keys,
                  values,
                  expect_keys,
                  expect_vals,
                  sum_agg(),
                  use_sort_option,
                  null_policy::INCLUDE);
};
do_test(force_use_sort_impl::YES);
do_test(force_use_sort_impl::NO);
```

Doing so, you can completely remove the two functions above this.

Reply:
Done